### ISSUE TITLE
GLIMPSE_chunk Segmentation fault (core dumped)

### ISSUE THREAD

**Issue by thierrygrange** [odelaneau/GLIMPSE#11]
Dear colleagues

When trying to run GLIMPSE_chunk it fails after a while (a few seconds to a few hundred seconds depending on input file size, the larger, the longer). The error message is Segmentation fault (core dumped). 

Any idea about the underlying cause and the possible solutions?

Thanks

Thierry

**Comment by SimoneRubinacci** [comment]
Hi Thierry,

It's a very general error, that can be caused from a wide range of problems. Might be a bug or even the data (even though, a message should be printed instead of a segfault). Can you post more details about your data and the commands you are using?

Best, 
Simone

**Comment by thierrygrange** [comment]
Dear Simone

We ran the command with two different inputs. One is as it is in the tutorial with the reference panel excluding the individual of interest, which is NA12878. In the whole workflow, the tutorial makes us prepare this reference panel excluding this individual and use it as reference. But on the other hand, we also used the input as the raw reference panel as a test since this will be the one we want to use for our analyses. Anyway, the outcome are the same. The only difference is the time it takes before the error message appears. Using the input with the modified panel takes around 6 sec and using the full reference panel, it takes around 350 sec.

Here are the commands and outputs.

Thank for your suggestions.

Best

Thierry

Input=1000GP.chr22.noNA12878.sites.vcf.gz

GLIMPSE_chunk --input reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz --region 22 --window-size 2000000 --buffer-size 200000 --output chunks.chr22.txt


[GLIMPSE] Split chromosomes into chunks

  * Author        : Simone RUBINACCI & Olivier DELANEAU, University of Lausanne

  * Contact       : simone.rubinacci@unil.ch & olivier.delaneau@unil.ch

  * Version       : 1.0.0

  * Run date      : 28/05/2020 - 23:13:26


Files:

  * Input VCF      : [reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]

  * Chromosome     : [22]

  * Output file    : [chunks.chr22.txt]


Parameters:

  * Seed             : 15052011

  * Min. Window size : 2000000bp / 1000 variants

  * Min. Buffer size : 200000bp / 100 variants


Reading input files

  * Main      : [reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]

  * #variants = 0 (6.64s)


Splitting data into chunks and writting to [chunks.chr22.txt]

Segmentation fault (core dumped)


Input=ALL.chr22_GRCh38.genotypes.20170504.vcf.gz

 GLIMPSE_chunk --input ALL.chr22_GRCh38.genotypes.20170504.vcf.gz --region 22 --window-size 2000000 --buffer-size 200000 --output chunks.chr22.txt


[GLIMPSE] Split chromosomes into chunks

  * Author        : Simone RUBINACCI & Olivier DELANEAU, University of Lausanne

  * Contact       : simone.rubinacci@unil.ch & olivier.delaneau@unil.ch

  * Version       : 1.0.0

  * Run date      : 28/05/2020 - 22:16:31


Files:

  * Input VCF      : [ALL.chr22_GRCh38.genotypes.20170504.vcf.gz]

  * Chromosome     : [22]

  * Output file    : [chunks.chr22.txt]


Parameters:

  * Seed             : 15052011

  * Min. Window size : 2000000bp / 1000 variants

  * Min. Buffer size : 200000bp / 100 variants


Reading input files

  * Main      : [ALL.chr22_GRCh38.genotypes.20170504.vcf.gz]

  * #variants = 0 (350.21s)


Splitting data into chunks and writting to [chunks.chr22.txt]

Segmentation fault (core dumped)

**Comment by SimoneRubinacci** [comment]
Interesting, seems like a problem with the chromosome name since it reports #variants=0 in the region. I am not sure to understand what's happening, but maybe it's simply the --region option. This must perfectly fit with chr chromosome field in the vcf file. Can you try using "chr22" instead of "22"? 
22 is ok for build 37 data, but from what I understand you are using b38 data (the new tutorial), in this case it's used a different chromosome name. 

`GLIMPSE_chunk --input reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz --region **chr22** --window-size 2000000 --buffer-size 200000 --output chunks.chr22.txt`

Let me know if it works.

Best wishes,
Simone

**Comment by SimoneRubinacci** [comment]
I've now added an error in the case the number of markers in the region is zero. So you shouldn't get a segfault anymore.

**Comment by thierrygrange** [comment]
Dear Simone

Thanks for your help. Now if I run the command you indicated after updating I have the new error message, but I still do not understand what's wrong:
GLIMPSE_chunk --input reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz --region chr22 --window-size 2000000 --buffer-size 200000 --output chunks.chr22.txt

[GLIMPSE] Split chromosomes into chunks
  * Author        : Simone RUBINACCI & Olivier DELANEAU, University of Lausanne
  * Contact       : simone.rubinacci@unil.ch & olivier.delaneau@unil.ch
  * Version       : 1.0.0
  * Run date      : 31/05/2020 - 22:06:59

Files:
  * Input VCF      : [reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]
  * Chromosome     : [chr22]
  * Output file    : [chunks.chr22.txt]

Parameters:
  * Seed             : 15052011
  * Min. Window size : 2000000bp / 1000 variants
  * Min. Buffer size : 200000bp / 100 variants

Reading input files
  * Main      : [reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]
  * #variants = 0 (6.52s)

ERROR: No markers in region: chr22. Check chromosome name and start/end positions.

However, there should be markers in the file since if I run:
bcftools query -f '%CHROM  %POS  %REF  %ALT{0}\n' reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz | head

The output is:

chr22  10874444  C  T
chr22  11122151  G  A
chr22  11122417  G  T
chr22  11123361  C  A
chr22  11123683  T  C
chr22  11123736  C  T
chr22  11124625  G  T
chr22  11125238  T  C
chr22  11125296  A  T
chr22  11125438  T  C

Any idea of what could happen?

Best

Thierry

**Comment by SimoneRubinacci** [comment]
Very strange indeed. I can't really tell what's wrong from the log you provided. 
I've just re-run the scripts from steps1 to step4 in the tutorial and GLIMPSE_chunk runs as expected to me. 

`./step4_script_chunk.sh 

[GLIMPSE] Split chromosomes into chunks
  [...]
Files:
  * Input VCF      : [reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]
  * Chromosome     : [chr22]
  * Output file    : [chunks.chr22.txt]
[...]
Reading input files
  * Main      : [reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]
  * #variants = 1053373 (2.28s)

Splitting data into chunks and writting to [chunks.chr22.txt]
  * Internal window [chr22:10874444-50805809] / L=39931366bp / C=1053373
[...]
`
If I run "bcftools query/view" command on the sites.vcf.gz file, I get the same output as you. Since we use htslib, I can't see why bcftools does find records and GLIMPSE doesn't in your case. Can you please clean everything (running the script: step9_script_cleanup.sh) and re-run step1 to step4 (will take just few minutes)? One possible explanation I have is that the tabix-index you are using is not updated with this version of the data, for some reason.

Basically if the following commands outputs the same records and the the command you run, GLIMPSE must also retrieve the same records, since it's based on the same library:
`bcftools view -r chr22 reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz | bcftools query -f '%CHROM %POS %REF %ALT{0}\n' | head`

Please clean everything, re-run the scripts step1 to step4 and let me know if the problem is still there.

**Comment by thierrygrange** [comment]
Dear Simone

I followed your suggestion

The end of step 3 fails as follows:
bcftools mpileup -f ${REFGEN} -I -E -a 'FORMAT/DP' -T ${VCF} -r chr22 ${BAM} -Ou | bcftools call -Aim -C alleles -T ${TSV} -Oz -o ${OUT} 
Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid
[E::fai_build3_core] Failed to open the file reference-genome/hs38DH.chr22.fa.gz
Failed to read from standard input: unknown file type

The file is there, but maybe it has a problem? Or the problem is with the hs38DH.chr22.fa.gz.fai file?

Best

Thierry

**Comment by SimoneRubinacci** [comment]
Hi Thierry,
Which version of bcftools are you using? I have tested the dataset with bcftools 1.10 and there are no problems for me. 

Best wishes,
Simone

**Comment by thierrygrange** [comment]
Dear Simone

I am also using bcftools 1.10. I r
estarted from scratch, this time by really your scripts exactly, not by copying and pasting the commands like I did before. This time step 3 worked well without problem. 

Then I launched step4, the glimpse_chunk step. And it failed with the same last error message:

./step4_script_chunk.sh 

[GLIMPSE] Split chromosomes into chunks
  * Author        : Simone RUBINACCI & Olivier DELANEAU, University of Lausanne
  * Contact       : simone.rubinacci@unil.ch & olivier.delaneau@unil.ch
  * Version       : 1.0.0
  * Run date      : 02/06/2020 - 23:00:59

Files:
  * Input VCF      : [reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]
  * Chromosome     : [chr22]
  * Output file    : [chunks.chr22.txt]

Parameters:
  * Seed             : 15052011
  * Min. Window size : 2000000bp / 1000 variants
  * Min. Buffer size : 200000bp / 100 variants

Reading input files
  * Main      : [reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]
  * #variants = 0 (6.43s)

ERROR: No markers in region: chr22. Check chromosome name and start/end positions.

Thanks for your help in figuring out why it failed. I am puzzled.

Best

Thierry

**Comment by SimoneRubinacci** [comment]
Hi Thierry,

I really don't undertand why it's not working for you. Can you share your input files with me? Should be:
`reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz`
`reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz.csi`

Best wishes, 
Simone

**Comment by thierrygrange** [comment]
Dear Simone

I attach here the index only. The other file is too big for github and my mail and can be found at https://mycore.core-cloud.net/index.php/s/6wi14Xp22mgIFMv

Thanks for your help.

Thierry

**Comment by SimoneRubinacci** [comment]
Hi,

Thanks for this.
I've just downloaded your vcf.gz file from the link, indexed it using bcftools (`bcftools index -f 1000GP.chr22.noNA12878.sites.vcf.gz`) and run GLIMPSE_chunk: 
`GLIMPSE_chunk --input 1000GP.chr22.noNA12878.sites.vcf.gz --region chr22 --window-size 2000000 --buffer-size 200000 --output chunks.chr22.txt`

It all worked as expected to me, and I can generate the chunks. I couldn't find your index file (even though you said you sent it?) so I re-generated it using bcftools. I invite you to do the same, since the data is fine and the program should work as long as the index is computed correctly.

I am happy to share with you a static binary of GLIMPSE_chunk if that might help.
Best wishes

Simone

**Comment by thierrygrange** [comment]
Hi Simone

I did all what you suggest, and still same problem :

bcftools index -f reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz
GLIMPSE_chunk --input reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz --region chr22 --window-size 2000000 --buffer-size 200000 --output chunks.chr22.txt

[GLIMPSE] Split chromosomes into chunks
  * Author        : Simone RUBINACCI & Olivier DELANEAU, University of Lausanne
  * Contact       : simone.rubinacci@unil.ch & olivier.delaneau@unil.ch
  * Version       : 1.0.0
  * Run date      : 03/06/2020 - 19:04:52

Files:
  * Input VCF      : [reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]
  * Chromosome     : [chr22]
  * Output file    : [chunks.chr22.txt]

Parameters:
  * Seed             : 15052011
  * Min. Window size : 2000000bp / 1000 variants
  * Min. Buffer size : 200000bp / 100 variants

Reading input files
  * Main      : [reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]
  * #variants = 0 (6.60s)

ERROR: No markers in region: chr22. Check chromosome name and start/end positions.

Let’s try with a static binary, at least we will test if the compilation went wrong or if I have other sort of incompatibility on my computer.

Thanks

Thierry

**Comment by SimoneRubinacci** [comment]
Dear Thierry,

A static binary is available in the GLIMPSE/static_bins folder. It's not the latest version of the chunking program, but it should be okay for your purpose. 

Let me know if it works. Will update the binaries if that would be helpful for you,

Best,
Simone

**Comment by thierrygrange** [comment]
Dear Simone

It worked with the static_bins. Here is the output. I would be happy if you can update the static binaries and I would work with them.

Thanks

Thierry

static_bins/GLIMPSE_chunk --input test/reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz --region chr22 --window-size 2000000 --buffer-size 200000 --output chunks.chr22.txt

[GLIMPSE] Split chromosomes into chunks
  * Author        : Simone RUBINACCI & Olivier DELANEAU, University of Lausanne
  * Contact       : simone.rubinacci@unil.ch & olivier.delaneau@unil.ch
  * Version       : 1.0.0
  * Run date      : 18/06/2020 - 20:23:05

Files:
  * Input VCF      : [test/reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]
  * Chromosome     : [chr22]
  * Output file    : [chunks.chr22.txt]

Parameters:
  * Seed             : 15052011
  * Min. Window size : 2000000bp / 1000 variants
  * Min. Buffer size : 200000bp / 100 variants

Reading input files
  * Main      : [test/reference_panel/1000GP.chr22.noNA12878.sites.vcf.gz]
  * #variants = 1053373 (4.44s)

Splitting data into chunks and writting to [chunks.chr22.txt]
  * Internal window [chr22:10874444-50805809] / L=39931366bp / C=1053373
  * Internal window [chr22:10874444-34544246] / L=23669803bp / C=526687
  * Internal window [chr22:10874444-25330866] / L=14456423bp / C=263344
  * Internal window [chr22:10874444-20904452] / L=10030009bp / C=131673
  * Terminal window [0] -buffer:[chr22:10874444-18389964] / +buffer:[chr22:10874444-18590306] / L=7515521bp / C=65837
  * Terminal window [1] -buffer:[chr22:18390099-20904452] / +buffer:[chr22:18189926-21104458] / L=2514354bp / C=65836
  * Internal window [chr22:20904554-25330866] / L=4426313bp / C=131671
  * Terminal window [2] -buffer:[chr22:20904554-23229058] / +buffer:[chr22:20704512-23429156] / L=2324505bp / C=65836
  * Terminal window [3] -buffer:[chr22:23229095-25330866] / +buffer:[chr22:23029077-25530868] / L=2101772bp / C=65835
  * Internal window [chr22:25330929-34544246] / L=9213318bp / C=263343
  * Internal window [chr22:25330929-29951253] / L=4620325bp / C=131672
  * Terminal window [4] -buffer:[chr22:25330929-27433832] / +buffer:[chr22:25130928-27633833] / L=2102904bp / C=65837
  * Terminal window [5] -buffer:[chr22:27433849-29951253] / +buffer:[chr22:27233744-30151314] / L=2517405bp / C=65835
  * Internal window [chr22:29951262-34544246] / L=4592985bp / C=131671
  * Terminal window [6] -buffer:[chr22:29951262-32368241] / +buffer:[chr22:29751221-32568249] / L=2416980bp / C=65836
  * Terminal window [7] -buffer:[chr22:32368274-34544246] / +buffer:[chr22:32168271-34744248] / L=2175973bp / C=65835
  * Internal window [chr22:34544260-50805809] / L=16261550bp / C=526686
  * Internal window [chr22:34544260-43439299] / L=8895040bp / C=263344
  * Internal window [chr22:34544260-38883501] / L=4339242bp / C=131673
  * Terminal window [8] -buffer:[chr22:34544260-36693546] / +buffer:[chr22:34344250-36893568] / L=2149287bp / C=65837
  * Terminal window [9] -buffer:[chr22:36693559-38883501] / +buffer:[chr22:36493559-39083515] / L=2189943bp / C=65836
  * Internal window [chr22:38883510-43439299] / L=4555790bp / C=131671
  * Terminal window [10] -buffer:[chr22:38883510-41268483] / +buffer:[chr22:38683497-41468485] / L=2384974bp / C=65836
  * Terminal window [11] -buffer:[chr22:41268509-43439299] / +buffer:[chr22:41068439-43639393] / L=2170791bp / C=65835
  * Internal window [chr22:43439310-50805809] / L=7366500bp / C=263342
  * Terminal window [12] -buffer:[chr22:43439310-47269130] / +buffer:[chr22:43239268-47469184] / L=3829821bp / C=131672
  * Terminal window [13] -buffer:[chr22:47269166-50805809] / +buffer:[chr22:47069138-50805809] / L=3536644bp / C=131670
  * #chunks = 14

Total running time = 4 seconds

________________________________
De : Simone Rubinacci [notifications@github.com]
Envoyé : lundi 15 juin 2020 13:41
À : odelaneau/GLIMPSE
Cc : GRANGE Thierry; Author
Objet : Re: [odelaneau/GLIMPSE] GLIMPSE_chunk Segmentation fault (core dumped) (#11)


Dear Thierry,

A static binary is available in the GLIMPSE/static_bins folder. It's not the latest version of the chunking program, but it should be okay for your purpose.

Let me know if it works. Will update the binaries if that would be helpful for you,

Best,
Simone

**Comment by SimoneRubinacci** [comment]
Hi Thierry,

I am glad you fiinally got GLIMPSE to work!! In the long run I recommend to try to solve your compilation problems. 
Will produce the binaries for you tomorrow, and will send them to you by email. 
Hopefully soon there will be a new release so I will be easier to download a tarball.

Best wishes,

Simone

**Comment by thierrygrange** [comment]
Dear Simone

Thanks.

The difficulty to solve the compilation problem is that there was no error message during compilation, so it will be difficult to troubleshoot which library was faulty during compilation. i suspect that there is something in my computer configuration that cause the problem, but figuring out which one is difficult without error message and it may be linked to the requirements of a previously installed software, so I felt really stuck here. In the long run, hopefully you will make a conda or a docker package.

We are looking forward to explore your very appealing software with our data.

Best

Thierry
________________________________
De : Simone Rubinacci [notifications@github.com]
Envoyé : jeudi 18 juin 2020 21:17
À : odelaneau/GLIMPSE
Cc : GRANGE Thierry; Author
Objet : Re: [odelaneau/GLIMPSE] GLIMPSE_chunk Segmentation fault (core dumped) (#11)


Hi Thierry,

I am glad you fiinally got GLIMPSE to work!! In the long run I recommend to try to solve your compilation problems.
Will produce the binaries for you tomorrow, and will send them to you by email.
Hopefully soon there will be a new release so I will be easier to download a tarball.

Best wishes,

Simone

**Comment by thierrygrange** [comment]
Dear Simone

In the static_bins folder, there are currently only 4 GLIMPSE tools: GLIMPSE_chunk  GLIMPSE_ligate  GLIMPSE_phase  GLIMPSE_sample. I have only compiled version of GLIMPSE_impute GLIMPSE_concordance and GLIMPSE_snparray. Given the problem encountered with compilation as experienced with chunk, I anticipate problems when I will come to use these 3 other tools. Would it be possible to obtain static binaries for these tools as well?

Thanks

Best

Thierry
________________________________
De : GRANGE Thierry
Envoyé : jeudi 18 juin 2020 21:27
À : odelaneau/GLIMPSE; odelaneau/GLIMPSE
Cc : Author
Objet : RE:[odelaneau/GLIMPSE] GLIMPSE_chunk Segmentation fault (core dumped) (#11)

Dear Simone

Thanks.

The difficulty to solve the compilation problem is that there was no error message during compilation, so it will be difficult to troubleshoot which library was faulty during compilation. i suspect that there is something in my computer configuration that cause the problem, but figuring out which one is difficult without error message and it may be linked to the requirements of a previously installed software, so I felt really stuck here. In the long run, hopefully you will make a conda or a docker package.

We are looking forward to explore your very appealing software with our data.

Best

Thierry
________________________________
De : Simone Rubinacci [notifications@github.com]
Envoyé : jeudi 18 juin 2020 21:17
À : odelaneau/GLIMPSE
Cc : GRANGE Thierry; Author
Objet : Re: [odelaneau/GLIMPSE] GLIMPSE_chunk Segmentation fault (core dumped) (#11)


Hi Thierry,

I am glad you fiinally got GLIMPSE to work!! In the long run I recommend to try to solve your compilation problems.
Will produce the binaries for you tomorrow, and will send them to you by email.
Hopefully soon there will be a new release so I will be easier to download a tarball.

Best wishes,

Simone

**Comment by SimoneRubinacci** [comment]
Hi Thierry,

Do not worry about GLIMPSE_impute and GLIMPSE_snparray, the first is not maintained anymore and the second is completely undocumented. It's really meant for developers rather than users.
Will send GLIMPSE_concordance though, since it's now part of the tutorial.
Will also consider of releasing a docker package of the next release.

Best wishes,
Simone

**Comment by thierrygrange** [comment]
Dear Simone

A follow-up question on a different Glimpse topic.

To use Glimpse with our data, we have followed your instructions for SNP calling with bcftools with our data, which are ancient genomics data with low to medium coverage. With certain combinations of samples and chromosomes, SNP calling failed. This is occurring in more than half of the combinations. Exploring in details the underlying cause, slicing the script in various substeps and looking in details at the output files and the error messages, we could narrow it down to the bcftools call step when working with the multiallelic caller. If we use the original bcftools consensus caller instead (-c instead of -m), it works fine with our files, and we have apparently been able to do imputation and phasing of our test files.

We would like your opinions and feedbacks about us using this consensus calling mode to do imputation and phasing using Glimpse.

We thank you for your insights and for sharing your experience.

With best regards

Thierry Grange
________________________________
De : Simone Rubinacci [notifications@github.com]
Envoyé : vendredi 19 juin 2020 18:49
À : odelaneau/GLIMPSE
Cc : GRANGE Thierry; Author
Objet : Re: [odelaneau/GLIMPSE] GLIMPSE_chunk Segmentation fault (core dumped) (#11)


Hi Thierry,

Do not worry about GLIMPSE_impute and GLIMPSE_snparray, the first is not maintained anymore and the second is completely undocumented. It's really meant for developers rather than users.
Will send GLIMPSE_concordance though, since it's now part of the tutorial.
Will also consider of releasing a docker package of the next release.

Best wishes,
Simone

**Comment by odelaneau** [comment]
Just a question before going deeper into the details. Do you also try to call small indels with bcftools? From our experience, our pipeline does not support small indels yet (we had similar issues before).

**Comment by thierrygrange** [comment]
Yes, we did use the bcftools mpileup with the -I option to skip indels. Actually, we did follow all your recommendations for SNP calling, and the failures we observed, that were not systematic and occurring only with specific combinations of chromosomes and samples, were not occurring simply by replacing during the bcftools call step the -m -C alleles options by -c
All other parameters used were exactly like you suggests in the tutorial
Thanks for your help and insights
________________________________
De : Olivier Delaneau [notifications@github.com]
Envoyé : jeudi 27 août 2020 09:38
À : odelaneau/GLIMPSE
Cc : GRANGE Thierry; Author
Objet : Re: [odelaneau/GLIMPSE] GLIMPSE_chunk Segmentation fault (core dumped) (#11)


Just a question before going deeper into the details. Do you also try to call small indels with bcftools? From our experience, our pipeline does not support small indels yet (we had similar issues before).
